creusot/src/translation/function.rs

```diff
@@ -928,12 +973,107 @@ impl<'tcx> TranslationCtx<'tcx> {
         let mut csubst =
             closure_capture_subst(self, def_id, subst, true, Some(self_.clone()), self_);
 
-        let mut postcondition = postcondition.clone();
+        let mut postcondition = postcond(ClosureKind::FnOnce);
         csubst.visit_mut_term(&mut postcondition);
         contracts.postcond_once = Some(postcondition);
 
         contracts
     }
+
+    fn inferred_precondition_term(
+        &self,
+        def_id: DefId,
+        args: GenericArgsRef<'tcx>,
+        kind: ClosureKind,
+        closure_env: Term<'tcx>,
+        mut closure_args: Vec<Term<'tcx>>,
+        span: Span,
+    ) -> Term<'tcx> {
+        let env_ty = closure_env.ty;
+
+        let bor_self = Term::var(
+            Symbol::intern("__bor_self"),
+            Ty::new_ref(self.tcx, self.tcx.lifetimes.re_erased, env_ty, Mutability::Mut),
+        );
+        // Based on the underlying kind of closure we may need to wrap the call in a quantifier (at least for FnMut ones)
+        match kind {
+            ClosureKind::Fn | ClosureKind::FnOnce => {
+                closure_args.insert(0, closure_env.clone());
+
+                Term {
+                    kind: TermKind::Precondition { item: def_id, args, params: closure_args },
+                    ty: self.types.bool,
+                    span,
+                }
+            }
+            ClosureKind::FnMut => {
+                closure_args.insert(0, bor_self.clone());
+                let base = Term {
+                    kind: TermKind::Precondition { item: def_id, args, params: closure_args },
+                    ty: self.types.bool,
+                    span,
+                };
+                (bor_self.clone().cur().bin_op(self.tcx, BinOp::Eq, closure_env))
+                    .implies(base)
+                    .forall(self.tcx, (Symbol::intern("__bor_self"), env_ty))
+            }
+        }
+    }
+
+    /// Infers the `postcond_kind` version of the postcondition predicate for the provided closure.
+    fn inferred_postcondition_term(
+        &self,
+        postcond_kind: ClosureKind,
+        def_id: DefId,
+        args: GenericArgsRef<'tcx>,
+        closure_kind: ClosureKind,
+        closure_env: Term<'tcx>,
+        mut closure_args: Vec<Term<'tcx>>,
+        span: Span,
+    ) -> Term<'tcx> {
+        let env_ty = closure_env.ty;
+
+        match closure_kind {
+            ClosureKind::Fn | ClosureKind::FnOnce => {
+                closure_args.insert(0, closure_env.clone());
+
+                let base = Term {
+                    kind: TermKind::Postcondition { item: def_id, args, params: closure_args },
+                    ty: self.types.bool,
+                    span,
+                };
+                base
```

Comment:
I don't understand all the details, but I find quite suspicious that there is no special case here when `postcond_kind` is `FnMut`.

More generally, I would have expected to only generate the term when `closure_kind` is equal to `postcond_kind`, and rely on the various compatibility laws to let the provers deduce the other cases.

Reply:
> this fact being provided by the compatibility law.

Well, then, why not relying fully on the compatibility law for the whole post-condition?